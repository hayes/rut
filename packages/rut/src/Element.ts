import { ReactTestInstance, act } from 'react-test-renderer';
import { Args, UnknownProps } from './types';
import Queryable from './Queryable';

export default class Element<Props = UnknownProps> extends Queryable {
  private element: ReactTestInstance;

  constructor(element: ReactTestInstance) {
    super();

    this.element = element;
  }

  children(): (string | Element)[] {
    return this.testInstance().children.map(child => {
      if (typeof child === 'string') {
        return child;
      }

      return new Element(child);
    });
  }

  emit<K extends keyof Props>(name: K, ...args: Args<Props[K]>): unknown {
    const prop = this.prop(name);

    if (!prop) {
      throw new Error(`${name} does not exist.`);
    } else if (typeof prop !== 'function') {
      throw new TypeError(`${name} is not a function.`);
    }

    let value;

    act(() => {
      value = prop(...args);
    });

    return value;
  }

  prop<K extends keyof Props>(name: K): Props[K] | undefined {
    return this.testInstance().props[name as string];
  }

  props(): Props {
    return this.testInstance().props as Props;
  }

  ref<T>(name?: string): T | null {
    const inst = this.testInstance().instance;
    const fiber = this.testInstance()._fiber;

    if (!name || fiber.ref) {
      return fiber.ref;
    }

    // Callback refs
    if (name in inst) {
      return inst[name];
    }

    // String refs
    if (inst.refs[name]) {
      return inst.refs[name];
    }

    return null;
  }

  type(): React.ElementType {
    return this.testInstance().type;
  }

  protected testInstance() {
    return this.element;
  }
}
