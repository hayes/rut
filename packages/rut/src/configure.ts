import { RendererOptions } from './types';

export const globalOptions: RendererOptions = {};

export function configure(options: RendererOptions) {
  Object.assign(globalOptions, options);
}
