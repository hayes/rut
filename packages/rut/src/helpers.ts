import React from 'react';

export function getTypeName(type: React.ElementType): string {
  if (typeof type === 'function') {
    return type.displayName || type.name;
  }

  return type || 'UNKNOWN';
}
