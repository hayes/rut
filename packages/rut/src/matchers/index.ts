import toBeChecked from './toBeChecked';
import toBeDisabled from './toBeDisabled';
import toBeElementType from './toBeElementType';
import toBeNodeType from './toBeNodeType';
import toContainNode from './toContainNode';
import toHaveKey from './toHaveKey';
import toHaveProp from './toHaveProp';
import toHaveValue from './toHaveValue';
import toRenderChildren from './toRenderChildren';

export default {
  toBeChecked,
  toBeDisabled,
  toBeElementType,
  toBeNodeType,
  toContainNode,
  toHaveKey,
  toHaveProp,
  toHaveValue,
  toRenderChildren,
};
