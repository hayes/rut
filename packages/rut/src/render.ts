import React from 'react';
import { act } from 'react-test-renderer';
import Renderer from './Renderer';
import { RendererOptions } from './types';

export default function render<Props>(
  element: React.ReactElement<Props>,
  options?: RendererOptions,
): Renderer<Props> {
  let renderer: Renderer<Props>;

  act(() => {
    renderer = new Renderer(element, options);
  });

  return renderer!;
}
