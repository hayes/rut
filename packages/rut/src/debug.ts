import util from 'util';
import React from 'react';
import { ReactTestRendererTree } from 'react-test-renderer';
import { getTypeName, shallowEqual } from './helpers';

export function formatValue(value: unknown): string {
  const typeOf = typeof value;

  if (typeOf === 'string') {
    return `"${value}"`;
  } else if (
    typeOf === 'number' ||
    typeOf === 'boolean' ||
    value === null ||
    value instanceof RegExp
  ) {
    return String(value);
  }

  return `\`${String(value)}\``;
}

function toArray<T>(value?: null | T | T[]): T[] {
  if (!value) {
    return [];
  }

  return Array.isArray(value) ? value : [value];
}

function findMatchingNodeProp(
  nodes: ReactTestRendererTree[],
  element: React.ReactElement,
): ReactTestRendererTree | null {
  return (
    nodes.find(node => {
      if (node.type === element.type && shallowEqual(node.props, element.props)) {
        return node;
      }

      if (node.rendered) {
        return findMatchingNodeProp(toArray(node.rendered), element);
      }

      return undefined;
    }) || null
  );
}

function sortAndFormatProps(
  names: string[],
  props: ReactTestRendererTree['props'],
  nodes: ReactTestRendererTree[],
): string[] {
  const output: string[] = [];

  names.sort().forEach(name => {
    const value = props[name];
    const typeOf = typeof value;

    if (value === undefined) {
      return;
    } else if (value === true) {
      output.push(name);

      return;
    }

    let propValue;

    if (typeOf === 'string') {
      propValue = `"${value}"`;
    } else if (typeOf === 'function') {
      propValue = `${value.name || 'func'}()`;
    } else if (typeOf === 'object') {
      console.log({ name, typeOf, value });

      // Element
      if (React.isValidElement(value)) {
        // eslint-disable-next-line
        propValue = debug(findMatchingNodeProp(nodes, value));
        // Ref
      } else if ('current' in value) {
        propValue = getTypeName(value.current);
        // Arrays, objects, maps, sets
      } else {
        propValue = util
          .inspect(value, {
            depth: 1,
            maxArrayLength: 5,
          })
          .replace(/\{ /gu, '{')
          .replace(/\[ /gu, '[')
          .replace(/ \}/gu, '}')
          .replace(/ \]/gu, ']');
      }
    } else {
      propValue = formatValue(value);
    }

    if (propValue.startsWith('"')) {
      output.push(`${name}=${propValue}`);
    } else {
      output.push(`${name}={${propValue}}`);
    }
  });

  return output;
}

function formatProps(
  props: ReactTestRendererTree['props'],
  nodes: ReactTestRendererTree[],
): string {
  if (!props || typeof props !== 'object') {
    return '';
  }

  const all: string[] = [];
  const truthies: string[] = [];
  const handlers: string[] = [];

  Object.entries(props).forEach(([key, value]) => {
    if (key === 'children') {
      return;
    }

    if (value === true) {
      truthies.push(key);
    } else if (key.startsWith('on') && typeof value === 'function') {
      handlers.push(key);
    } else {
      all.push(key);
    }
  });

  const output = [
    ...sortAndFormatProps(truthies, props, nodes),
    ...sortAndFormatProps(all, props, nodes),
    ...sortAndFormatProps(handlers, props, nodes),
  ];

  return output.length === 0 ? '' : ` ${output.join(' ')}`;
}

function isAllTextNodes(nodes: unknown[]): boolean {
  return nodes.every(node => typeof node === 'string');
}

export default function debug(
  node: ReactTestRendererTree | string | null,
  depth: number = 0,
): string {
  if (!node) {
    return '';
  }

  const indent = '  '.repeat(depth);

  if (typeof node === 'string') {
    return `${indent}${node}`;
  }

  const nodes: ReactTestRendererTree[] = toArray(node.rendered);
  const name = getTypeName(node.type);
  const props = formatProps(node.props, nodes);

  if (nodes.length === 0) {
    return `${indent}<${name}${props} />`;
  }

  let children = '';

  // Inline if only text with no props
  if (isAllTextNodes(nodes) && props === '') {
    children = nodes.join('');
    // Otherwise continue nested indentation
  } else {
    children = `\n${nodes.map(child => debug(child, depth + 1)).join('\n')}\n${indent}`;
  }

  return `${indent}<${name}${props}>${children}</${name}>`;
}
