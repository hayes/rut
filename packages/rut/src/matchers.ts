/* eslint-disable no-magic-numbers, no-underscore-dangle */

import React from 'react';
import Element from './Element';
import { getElementTypeName } from './helpers';
import { MatchResult, NodeType } from './types';

function checkIsRutElement(matcher: string, value: unknown) {
  if (value instanceof Element) {
    return;
  }

  if (
    typeof value === 'object' &&
    value !== null &&
    Object.getPrototypeOf(value).constructor.name === 'Element' &&
    Object.getPrototypeOf(Object.getPrototypeOf(value)).constructor.name === 'Queryable'
  ) {
    return;
  }

  throw new Error(`${matcher}: Expected a Rut \`Element\`.`);
}

/**
 * Check that an `Element` is a valid React element type.
 * Accepts either a class or function component, or the name of a host component (HTML tag).
 */
export function toBeElementType(received: Element, type: React.ElementType): MatchResult {
  checkIsRutElement('toBeElementType', received);

  const expectedName = getElementTypeName(type);

  return {
    message: `expected \`${received}\` to be a \`${expectedName}\``,
    notMessage: `expected \`${received}\` not to be a \`${expectedName}\``,
    passed: received.type() === type,
  };
}

// Keep in sync with React upstream!
// https://github.com/facebook/react/blob/master/packages/shared/ReactWorkTags.js
const nodeTypeMap: { [K in NodeType]: number | number[] } = {
  'class-component': 1,
  // 'context-consumer': 9,
  // 'context-provider': 10,
  'forward-ref': 11,
  fragment: [7, 18],
  'function-component': 0,
  'host-component': 5,
  'indeterminate-component': 2,
  lazy: 16,
  memo: [14, 15],
  mode: 8,
  portal: 4,
  profiler: 12,
  root: 3,
  suspense: [13, 19],
  text: 6,
};

/**
 * Check that an `Element` is a specific type of React node.
 * React nodes are based on React fiber work tags.
 */
export function toBeNodeType(received: Element, type: NodeType): MatchResult {
  checkIsRutElement('toBeNodeType', received);

  // @ts-ignore Allow access for matcher
  const fiberTag = received.testInstance()._fiber.tag;
  const nodeToTag = nodeTypeMap[type];

  if (nodeToTag === undefined) {
    throw new Error(
      `toBeNodeType: Invalid node type "${type}", expected one of: ${Object.keys(nodeTypeMap).join(
        ', ',
      )}`,
    );
  }

  return {
    message: `expected \`${received}\` to be a "${type}"`,
    notMessage: `expected \`${received}\` not to be a "${type}"`,
    passed: Array.isArray(nodeToTag) ? nodeToTag.includes(fiberTag) : nodeToTag === fiberTag,
  };
}

export function toContainNode(received: Element, node: NonNullable<React.ReactNode>): MatchResult {
  checkIsRutElement('toContainNode', received);

  const results = received.query(
    (testNode, fiberNode) => {
      if (React.isValidElement(node)) {
        return (
          testNode.type === node.type && testNode.props === node.props && fiberNode.key === node.key
        );
      }

      // RTR doesn't run the predicate on non-element nodes (like strings and numbers),
      // so we need to query the parent by traversing the children.
      // https://github.com/facebook/react/blob/master/packages/react-test-renderer/src/ReactTestRenderer.js#L388
      if (testNode.children.includes(String(node))) {
        return true;
      }

      return false;
    },
    { deep: false },
  );

  return {
    message: `expected \`${received}\` to contain node TODO`,
    notMessage: `expected \`${received}\` not to contain node TODO`,
    passed: results.length > 0,
  };
}

export function toRenderChildren(received: Element): MatchResult {
  checkIsRutElement('toRenderChildren', received);

  return {
    message: `expected \`${received}\` to render children`,
    notMessage: `expected \`${received}\` not to render children`,
    passed: received.children().length > 0,
  };
}
