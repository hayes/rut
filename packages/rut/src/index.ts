/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Element from './Element';
import Renderer from './Renderer';
import render from './render';

export * from './helpers';
export * from './types';

export { render, Element, Renderer };
