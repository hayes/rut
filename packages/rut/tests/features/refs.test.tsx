import React from 'react';
import { render } from '../../src/render';

describe('Refs', () => {
  const mock = { tagName: 'BUTTON' };

  describe('forwardRef', () => {
    const Button = React.forwardRef(
      ({ children }: { children: React.ReactNode }, ref: React.Ref<HTMLButtonElement>) => {
        return (
          <button type="button" ref={ref}>
            {children}
          </button>
        );
      },
    );

    it('resolves the ref from the inner component', () => {
      const ref = React.createRef<HTMLButtonElement>();
      const result = render(<Button ref={ref}>Child</Button>, {
        refs: { button: mock },
      });

      expect(result).toMatchSnapshot();
      expect(result.root.ref()).toBe(ref);
      expect(result.root.ref<typeof ref>()!.current).toBe(mock);
    });

    function withRef(WrappedComponent: React.ComponentType<{ children: React.ReactNode }>): any {
      class WithRef extends React.Component<{ forwardedRef?: React.Ref<unknown> | null }> {
        render() {
          const { forwardedRef, ...rest } = this.props;

          // @ts-ignore
          return <WrappedComponent ref={forwardedRef} {...rest} />;
        }
      }

      return React.forwardRef((props, ref) => {
        return <WithRef {...props} forwardedRef={ref} />;
      });
    }

    class InnerButton extends React.Component<{ children: React.ReactNode }> {
      render() {
        return <button type="button">{this.props.children}</button>;
      }
    }

    const HocButton = withRef(InnerButton);

    it('resolves the ref from the HOC', () => {
      const ref = React.createRef<HTMLButtonElement>();

      expect(ref.current).toBeNull();

      const result = render(<HocButton ref={ref}>Child</HocButton>);

      expect(result).toMatchSnapshot();
      expect(ref.current!.constructor.name).toBe('InnerButton');
    });
  });

  describe('createRef', () => {
    class Section extends React.Component<{ children: React.ReactNode }> {
      ref = React.createRef<HTMLDivElement>();

      render() {
        return <div ref={this.ref}>{this.props.children}</div>;
      }
    }

    it('resolves the ref', () => {
      const div = { tagName: 'DIV' };
      const result = render(<Section>Child</Section>, {
        refs: { div },
      });

      expect(result).toMatchSnapshot();
      expect(result.root.ref('ref')).toEqual({ current: div });
    });
  });

  describe('callback refs', () => {
    class Input extends React.Component<{ value: string }> {
      inputRef: HTMLInputElement | null = null;

      render() {
        return (
          <input
            value={this.props.value}
            ref={ref => {
              this.inputRef = ref;
            }}
          />
        );
      }
    }

    it('resolves the callback ref', () => {
      const input = { tagName: 'INPUT' };
      const result = render(<Input value="foo" />, {
        refs: { input },
      });

      expect(result).toMatchSnapshot();
      expect(result.root.ref('inputRef')).toBe(input);
    });
  });

  describe('string refs', () => {
    class Link extends React.Component<{ children: React.ReactNode; href: string }> {
      render() {
        return (
          // eslint-disable-next-line react/no-string-refs
          <a href={this.props.href} ref="link">
            {this.props.children}
          </a>
        );
      }
    }

    it('resolves the string ref', () => {
      const a = { tagName: 'A' };
      const result = render(<Link href="/">Child</Link>, {
        refs: { a },
      });

      expect(result).toMatchSnapshot();
      expect(result.root.ref('link')).toBe(a);
    });
  });
});
