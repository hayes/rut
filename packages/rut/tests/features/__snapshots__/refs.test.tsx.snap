// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Refs callback refs resolves the callback ref 1`] = `
"<Input value=\\"foo\\">
  <input value=\\"foo\\" />
</Input>"
`;

exports[`Refs createRef resolves the ref 1`] = `
"<Section>
  <div>Child</div>
</Section>"
`;

exports[`Refs forwardRef resolves the ref from the HOC 1`] = `
"<WithRef forwardedRef={InnerButton}>
  <InnerButton>
    <button type=\\"button\\">
      Child
    </button>
  </InnerButton>
</WithRef>"
`;

exports[`Refs forwardRef resolves the ref from the inner component 1`] = `
"<button type=\\"button\\">
  Child
</button>"
`;

exports[`Refs string refs resolves the string ref 1`] = `
"<Link href=\\"/\\">
  <a href=\\"/\\">
    Child
  </a>
</Link>"
`;
